BUGFIX: Restore `Neos.Neos:NodeUri`, `neos:link.node` and `LinkingService`

mhsdesign opened the pull request:
Resolves: https://github.com/neos/neos-development-collection/issues/4524
Related: https://github.com/neos/neos-development-collection/issues/4552

For linking to nodes in Neos (via `Neos.Neos:NodeUri`, `neos:link.node` or `neos:uri.node`) following logic will apply for Neos 9:

If a node is passed as string the base node is required.
Following string syntax is allowed:


- `/<Neos.Neos:Sites>/my-site/main`
- `some/relative/path`

Only allowed in Fluid: `node://my-node-identifier`

Deprecated syntax:

- `/sites/site/absolute/path`
- `~/site-relative/path`
- `~`

Not supported syntax:

- `./neos/info`
- `../foo/../../bar`


Changes:

- handle string node path syntax in `Neos.Neos:NodeUri` which requires the `baseNodeName` to be set (https://github.com/neos/neos-development-collection/issues/4524).
- `Neos.Neos:NodeUri` will like the fluid viewhelpers now also accept the `node://my-id` syntax
- as partial replacement of `\Neos\Neos\TYPO3CR\NeosNodeService::normalizePath` the new utility `\Neos\Neos\Utility\LegacyNodePathNormalizer::tryResolveLegacyPathSyntaxToAbsoluteNodePath` will handle the legacy path syntax
  - instead of reintroducing support of path traversal like `/../` via 8.3's `\Neos\ContentRepository\Domain\Utility\NodePaths::replaceRelativePathElements` we now correctly document that this syntax is not allowed and throw an descriptive error (previously `NodePath::fromString` would have thrown an error as the node name pattern is not supporting dots).
- the NodeViewHelper now correctly use the `baseNodeName`/ the documentNode to resolve string nodes
- for shortcuts the link NodeViewHelper (`neos:link.node`) will initialise the `linkedNode` variable as the shortcut itself and NOT resolve it. This is the same behaviour in Neos 8.3. Previously shortcut resolving was once earlier part of the uri builder meaning the `linkedNode` was the actual shortcut, but as Neos 8.3 didnt feature this, we wont need to fix this for fluid.
- the duplicated logic in the NodeViewHelper's `resolveNodeAddressFromPath` was extracted into `\Neos\Neos\Utility\NodeAddressNormalizer::resolveNodeAddressFromPath`
- the `LinkingService` was marked as deprecated and now uses the `LinkHelper` where possible
- the `LinkHelper` was simplified to fully work on psr `getScheme` and `getHost` instead of regex
- `\Neos\Neos\Fusion\Helper\LinkHelper::resolveNodeUri` was marked as deprecated as its not usable from within fusion, and it requires the `LinkingService`
- setting `node` to null in the fluid viewhelpers and fusion will not lead to link generation of the `baseNodeName`/ the documentNode but an exception
- the pattern `PATTERN_SUPPORTED_URIS` in convert uris was simplified as it was previously a union but the character set already matches uuids

Left open points:
- [x] naming of `NodeAddressNormalizer`
- [ ] find alternative for `LinkingService::createSiteUri`
- [x] why do we evaluate the hidden state in the `LinkingService`? We dont do it in the new uri builder
- [ ] remove `return $this->newLinkHelper->hasSupportedScheme($uri);` in the `LinkingService` and add rector migration?
- [x] add behat tests for `Neos.Neos:NodeUri`

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Introduce `LegacyNodePathNormalizer`

Converts legacy paths like "/absolute/path", "~/site-relative/path" and "~" to the corresponding
AbsoluteNodePath depending on the passed base node.
The following syntax is not implemented and handled here:
 - node://
 - /<Neos.Neos:Sites>/my-site/main
 - some/relative/path
Also while legacy and previously allowed, node path traversal like ./neos/info or ../foo/../../bar is not handled.

mhsdesign committed:
TASK: Introduce `NodeAddressNormalizer::resolveNodeAddressFromPath`

mhsdesign committed:
TASK: Simplify `NodeViewHelper`

Previously the `linkedNode` would contain the node the shortcut points to actually.
But as this is not the Neos 8.3 behaviour and introduces complexity that should be handled by the uri builder it will be removed again.

mhsdesign committed:
BUGFIX: Correctly use `baseNodeName` in `NodeViewHelper`'s

mhsdesign committed:
!!! TASK: Remove dead option `resolveShortcuts` from `neos:uri.node`

mhsdesign committed:
BUGFIX: Reintroduce relative node paths for `Neos.Neos:NodeUri`

mhsdesign committed:
TASK: Deprecate `LinkingService` and correct its logic

mhsdesign committed:
TASK: Fix `NodeAddressNormalizer` to handle empty paths

mhsdesign committed:
TASK: Improve exceptions

mhsdesign committed:
TASK: Simplify LinkHelper and deprecate `resolveNodeUri`

mhsdesign committed:
TASK: Use the `LinkHelper` als b/c layer in the `LinkingService`

mhsdesign committed:
TASK: Adjust phpstan types

mhsdesign committed:
TASK: Simplify regex of `PATTERN_SUPPORTED_URIS`

Previously it was a union of `[a-z0-9-]+` and a UUID pattern, but the character set already covers the UUID case

mhsdesign committed:
TASK: Document `NodeAddressNormalizer`

mhsdesign commented:
Another point is, should we actually introduce support again for these legacy path syntax when the most prominent places flow query, and many places in the configuration doesnt allow it? So having this runtime magic here for `Neos.Neos:NodeUri` might not be worth it, or at least not if we dont allow it again in the other places

- [Neos.Node.nearestContentCollection](https://github.com/neos/neos-development-collection/blob/931304f7ba76eccee6508c8fe63c13a813bcd239/Neos.Neos/Classes/Fusion/Helper/NodeHelper.php#L65)
  - supports new absolute syntax and relative
  - why do we support absolute paths here at all??
- [q(node).find](https://github.com/neos/neos-development-collection/blob/ec803ee9674f793426cdff944573c381135bdcc3/Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/FindOperation.php#L134)
  - supports new absolute syntax and relative
- [q(node).children](https://github.com/neos/neos-development-collection/blob/ec803ee9674f793426cdff944573c381135bdcc3/Neos.ContentRepository.NodeAccess/Classes/FlowQueryOperations/ChildrenOperation.php#L143)
  - supports relative syntax



see also: https://github.com/neos/rector/issues/24

bwaidelich commented on Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php:
```diff
@@ -67,8 +67,7 @@
 {
-    public const PATTERN_SUPPORTED_URIS
-        = '/(node|asset):\/\/([a-z0-9\-]+|([a-f0-9]){8}-([a-f0-9]){4}-([a-f0-9]){4}-([a-f0-9]){4}-([a-f0-9]){12})/';
+    private const PATTERN_SUPPORTED_URIS = '/(node|asset):\/\/([a-z0-9\-]+)/';
```
That pattern is definitely more correct – but it basically copies the regex from `NodeAggregateId::PATTERN` and there's a risk, that this gets out of sync (as it did _before_).
Probably out of scope for this PR but since you changed the regex, we could think about constructing the pattern from the original regex (along the lines of `/(node):\/\/' . NodeAggregateId::PATTERN...`

mhsdesign replied:
`NodeAggregateId::PATTERN` is private ... im fine with keeping it this way^^

bwaidelich reviewed (commented):
No +1 yet, because I failed to review this properly..
Just a little, half unrelated, comment.
Also I wonder: Is this behavior tested somehow?

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/restoreLinkingService

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/restoreLinkingService

mhsdesign committed:
TASK: Add Neos.Neos:NodeUri behat test for Neos9 :D

mhsdesign commented:
> Also I wonder: Is this behavior tested somehow?

was my last todo:) i introduced a test for 8.3, upmerged that and adjusted it here.
I didnt tests the exceptional cases yet because there is an absorbing throwable storage involved and i need a good plan.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/restoreLinkingService

mhsdesign committed:
TASK: Revert rename of `Neos.Neos/Tests/Behavior/behat.yml`

mhsdesign committed:
TASK: Remove `node://` protocol handling from `NodeAddressNormalizer`

That was also not in 8.3 supported in the linking service or Neos.Neos:NodeUri

mhsdesign committed:
TASK: Remove absolute nodepath feature from `nearestContentCollection`

This syntax is not required here

mhsdesign commented:
regarding https://github.com/neos/neos-development-collection/pull/4886#issuecomment-2213551160
i removed absolute node path functionality from [Neos.Node.nearestContentCollection](https://github.com/neos/neos-development-collection/blob/931304f7ba76eccee6508c8fe63c13a813bcd239/Neos.Neos/Classes/Fusion/Helper/NodeHelper.php#L65) and also tested if `.find()` worked in 8.3 with `~` and it doesnt. So we indeed dont have to support it here.
Further like in 8.3 only the view-helpers will interpret `node://` not the fusion node uri.

I also remember why i deprecated `~` with this change in the `LegacyNodePathNormalizer`: because nodePaths/ node names are deprecated in general and thus site relative paths are no use as well.

so definitely merge ready

// cc @kitsunet

mhsdesign committed:
TASK: Adjust doc comment

kitsunet commented:
I'll take a last look tomorrow/later and merge

kitsunet reviewed (approved):
The one comment about reusing the regex from basti seems like a good idea but by no means a blocker, thanks, this looks very ncie.

mhsdesign commented on Neos.Neos/Classes/Utility/NodeAddressNormalizer.php:
```diff
@@ -0,0 +1,79 @@
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+
+final class NodeAddressNormalizer
```
I think this is the wrong name still ... maybe `NodePathResolver::resolveNodeAddressFromPath` or `NodePathUtitlity::retrieveNodeAddressByPath`  idont know ...
or does *Normalizer fit the usecase here?

cc @kitsunet

kitsunet replied:
Yeah I definitely don't like Normalizer here, more like Converter or Resolver given that the method is named resolve...? 

Do we want this `@internal` to avoid people using it (as we probably don't want to keep this forever)?

mhsdesign replied:
Jup, added internal and renamed it to `NodePathResolver::resolveNodeAddressByPath` ... the normalise naming was stolen from 8.3 but doesnt fit anymore ^^

mhsdesign committed:
TASK: Rename `NodeAddressNormalizer` to `NodePathResolver`

mhsdesign committed:
TASK: Adjust deprecation notice

mhsdesign committed:
TASK: Fix hidden state evaluation in LinkingService and document new uri builder behaviour

This was accidentally wrongly migrated.

Regarding
> Why do we evaluate the hidden state here? We dont do it in the new uri builder.

We dont evaluate the hidden state in the new uri builder because we only have the node address at hand. The 8.3 logic

```
$action = $node->getContext()->getWorkspace()->isPublicWorkspace() && !$node->isHidden() ? 'show' : 'preview';
```

Ensured if youre routing to a hidden live node then a preview uri will be built. This is mostly unlikely going to happen and building the preview uri in this case is possibly not even the desired behaviour

One exception, in the Neos.Ui requires this behaviour: https://github.com/neos/neos-ui/pull/3867

bwaidelich commented on Neos.Neos/Classes/Utility/LegacyNodePathNormalizer.php:
```diff
@@ -0,0 +1,108 @@
+    public function tryResolveLegacyPathSyntaxToAbsoluteNodePath(
+        string $path,
+        Node $baseNode
+    ): ?AbsoluteNodePath {
```
I don't understand the return type. This can either return an `AbsoluteNodePath` or lead to an exception _or_ it returns `null`. What is the meaning of `null` and is that a good abstraction? Maybe we can at least add some `@return` doc comment explaining the ambiguity?

bwaidelich replied:
Ah I see, in the usages you do
```
tryResolveLegacyPathSyntaxToAbsoluteNodePath($path, ..) ?? $path
```
wouldn't it make things easier to have some
```php
$absoluteNodePath = AbsoluteNodePath::tryFromString($path);
if ($absoluteNodePath !== null) {
    return $absoluteNodePath;
}
```
at the beginning of this method and get rid of the nullable return type?

mhsdesign replied:
youre right to point this out:) I forgot to write down my line of thought here.

The idea is to separate the legacy behaviour from the more future proof, so we can later just remove the dependency to `tryResolveLegacyPathSyntaxToAbsoluteNodePath` ... calling `AbsoluteNodePath::tryFromString` would not be enough as the relative base node behaviour would also need to be handled ... 

its a little odd but its all `@internal` so maybe keep it this way?

Alternatively this could be the api:

```php
public function preprocessLegacyPathSyntax(
        string $path,
        Node $baseNode
    ): string;
```

mhsdesign replied:
its a little weird but for now well keep it okay its internal :)

mhsdesign replied:
thank you for the review though ❤️ though now its time to tag the beta:)

bwaidelich replied:
I'm absolutely fine with being overruled, but it's a bit demotivating if all my comments are just disregarded. You could have at least added the `@return` doc comment.
Anyways, not a big deal in this case of course

kitsunet replied:
you are right @bwaidelich we had a short call about it, and decided this is ultimately throw away code and we shouldn't bikeshed about it and instead get the beta out. The doc comment would definitely not hurt though. 
SORRY! ❤️wasn't meant to ignore you just to get the topic off of our (collective) heads, we discussed the other options but didn't see any single one that was much more pretty and therefore worth changing to.

bwaidelich replied:
no worries, no need to be sorry. But thanks for the explanation!

kitsunet commented on Neos.Neos/Classes/Utility/LegacyNodePathNormalizer.php:
```diff
@@ -0,0 +1,108 @@
+
+        if (AbsoluteNodePath::patternIsMatchedByString($path)) {
+            // not a legacy absolute node path
+            return null;
```
So couldn't this be `AbsoluteNodePath::tryFromString($path)`

kitsunet commented on Neos.Neos/Classes/Utility/LegacyNodePathNormalizer.php:
```diff
@@ -0,0 +1,108 @@
+        }
+
+        // not a legacy absolute node path
+        return null;
```
And then this an exception, as then we wouldn't need the null return value anymore...
